### ISSUE TITLE
[RFC] Async User Function for RPC

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#36071]
## Impact

Currently, every RPC request occupies an RPC thread on the server side until done. However, if there are nested RPC calls or other IO operations, we don't have any feature to allow user functions to yield on the server side. As a result, threads processing those requests would idle wait. The _Async User Function_ feature_ aims to fix this by allowing applications to mark a function as async. This function must return an `rpc.Future`, and RPC framework would install response processing/handling as a callback to that Future.

### API

```python
import torch.distributed.rpc as rpc

def my_add_2(a, b):
    return a + b

@rpc.async_function
def my_add_3(a, b, c):
    # final output future of this user function
    future_ret = rpc.Future()
    # nested RPC calls
    future_value = rpc.rpc_async(dst2, my_add_2, args=(a, b))
    # finish processing asynchronously when the nested RPC is done.
    future_value.add_callback(
        lambda ret: future_ret.mark_complete(ret + c)
    )
    return future_ret
    
rpc.rpc_sync(dst1, my_add_3, args=(1, 2, 3))
```

### New Concepts

1. `@rpc.async_function` decorator marks a function as an async function. 
    1. It gives the server side a hint that this is async user function, which can be done by registering the function name on the receiver, similar to how we handle `TorchScript` functions. 
    2. All async functions must return a future. 
2. `rpc.Future()` creates a future object which can be marked complete by the user function when the value is ready. This is not exactly a new concept, we just need to expose it and allow applications to create an empty Future.
3. `Future.add_callback` this is already available in C++, we need to expose it to Python.

### Nested Async User Functions

```python
import torch.distributed.rpc as rpc

def identity(a):
    return a

@rpc.async_function
def my_add_2(a, b):
    fut_ret = rpc.Future()
    rpc_async(dst, identity, args=(a, )).add_callback(
        lambda ret: fut_ret.mark_complete(ret + b)
    )
    return fut_ret
    

@rpc.async_function
def my_add_3(a, b, c):
    fut_ret = rpc.Future()
    rpc.rpc_async(dst, my_add_2, args=(a, b)).add_callback(
        lambda ret: fut_ret.mark_complete(ret + c)
    )
    return fut_ret
```

### Multiple Async Calls in One User Function

```python
def my_add_2(a, b):
    return a + b

@async_function
def my_add_3(a, b, c):
    fut_ret = rpc.Future()
    def bottom_half(x):
        fut2 = rpc.rpc_async(dst, my_add_2, args=(x, x))
        fut2.add_callback(
            lambda ret: fut_ret.mark_complete(ret)
        )
    
    rpc.rpc_async(dst, my_add_2, args=(a, b)).add_callback(bottom_half)
    return fut_ret
```

### Async RPC Fan Out

```python
def my_add_2(a, b):
    return a + b

@async_function
def my_add_4(a, b, c, d):
    fut_ret = rpc.Future()
    rets = []
    lock = threading.Lock()
    
    def barrier(x):
        flag = False
        with lock:
            rets.append(x)
            if len(rets) == 2:
                flag = True
        if flag:
            fut_ret.mark_complete(rets)
            
    rpc.rpc_async(dst2, my_add_2, args=(a, b)).add_callback(barrier)
    rpc.rpc_async(dst2, my_add_2, args=(c, d)).add_callback(barrier) 
    return fut_ret
```

### Batch RPC Requests

```python
lock = threading.Lock()
a_list = []
b_list = []
fut_rets = []
batch_size = 10

@async_function
def batch_add(a, b):
    flag = False
    fut_ret = rpc.Future()
    with lock:
        a_list.append(a)
        b_list.append(b)
        fut_rets.append(fut_ret)
        
        if len(a_list) == batch_size:
            tmp_a_list, a_list = a_list, []
            tmp_b_list, b_list = b_list, []
            tmp_fut_rets, fut_rets = fut_rets, []
            flag = True
            
    if flag:
        aa = torch.stack(a_list)
        bb = torch.stack(b_list)
        cc = aa + bb
        for i in range(len(fut_rets)):
            fut_rets[i].mark_complete(cc[i])
    
    return fut_ret
```

## Discussion

Q: why not using native Python `asyncio` API?
A: The `asyncio` proposal is started in Python 3.3. The `async`/`wait` syntax is first added in Python 3.5, and then becomes in Python 3.6. As RPC also aim to support earlier Python 3 releases, we cannot rely on a feature that is only stabilized for 3.6+.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @xush6528 @jjlilley @osalpekar

**Comment by jjlilley** [comment]
> Currently, every RPC request occupies an RPC thread on the server side until done.

fwiw, for  async TorchScript that uses jit.fork/wait underneath, this was actually addressed last week, but plumbing through interpreter runAsync() call into the request_callback_impl handler.

Definitely true about pure Python though.

**Comment by pritamdamania87** [comment]
> Future.add_callback this is already available in C++, we need to expose it to Python.

This is similar to what we do in C++, although I was wondering if it would be more natural for users to have a Future API similar to folly (thenValue() and thenError()). Something like this:

```
future_value.then_value(
        lambda ret: future_ret.mark_complete(ret + c)
    ).then_error(
       lambda error: print ('Received error {}'.format(error))
    )
```

It's a nice way for users to handle errors as well. I guess Jeremy's PR (https://github.com/pytorch/pytorch/pull/35849) for merging the Futures is kind of related here as well. Another option would be to expose an API like `has_error` (similar to https://github.com/pytorch/pytorch/pull/35849) for users to query whether or not there was an error.

I'm fine with either approach, thoughts @jjlilley @mrshenli ?

**Comment by jjlilley** [comment]
Not really sure which api would be best in Python here, but to add - the `folly::future` c++ apis aren't necessarily known for being easy to use well.

Sure, the initial demos look simple. But by the time you chain of 4-5 futures together, have errors skip over different part of the chain, do a collect loop, etc, and try to understand the underlying template meta-programming, there is some real complexity. 

Not sure how much of that complexity we want to emulate? If we don't need it all, a simple `addCallback()` might be sufficient (less is more?). But maybe I'm underplaying this?

**Comment by pritamdamania87** [comment]
> Not sure how much of that complexity we want to emulate? If we don't need it all, a simple addCallback() might be sufficient (less is more?). But maybe I'm underplaying this?

Makes sense, I guess we should then expose the `hasError` API to python so that users can check for errors within `add_callback`?

**Comment by jjlilley** [comment]
> > Not sure how much of that complexity we want to emulate? If we don't need it all, a simple addCallback() might be sufficient (less is more?). But maybe I'm underplaying this?
> 
> Makes sense, I guess we should then expose the `hasError` API to python so that users can check for errors within `add_callback`?

Yes, I think this would be a good idea. I suspect we will want a few other usability tweaks as well.

**Comment by lw** [comment]
Here is some references on what the API of futures is in the Python standard library:
- https://docs.python.org/3/library/concurrent.futures.html#future-objects
- https://docs.python.org/3/library/asyncio-future.html#future-object

I guess Python developers who are familiar with the concept of futures would expect to see such an API. They both have only a `add_done_callback` method, and the callback will receive the future object (not its value), which can thus also be in an error state by containing an exception. This seems to be similar to a folly::Try object.

The other standard API for dealing with this would be asyncio, as mentioned at the beginning. It can be simulated in pre-3.6 versions by using yield and yield from appropriately, but I think the result ends up being quite cryptic. Also, this type of asynchronicity through coroutines works well for single-threaded event loop-based libraries, and I'm not sure it's appropriate for RPC...

**Comment by wanchaol** [comment]
@mrshenli maybe a dumb question, the motivation of exposing Future constructor is mainly because we want to somehow have a way to return a single Future when it has multiple `rpc.async` functions in a `@rpc.async_function`? At least from the python concurrent package, it discourages/forbids people creating a future directly (see https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future)

**Comment by mrshenli** [comment]
@wanchaol 

> the motivation of exposing Future constructor is mainly because we want to somehow have a way to return a single Future when it has multiple

The original motivation is to allow user functions to yield the RPC thread on the server side when are waiting for IO or other events to unblock. Currently, there is no way to do this in user code. One use case would be multiple workers streaming requests to the same server, and the server would like to accumulate say 100 request and then stack them in one tensor to process to reduce the amortized CUDA kernel launching overhead. This is doable currently through either 1) let the request to block a thread on server or 2) let the server to launch another RPC to each worker to dispatch responses. But neither are preferable. A better solution would be creating a future response in the user code, and server can mark them as complete when the batched requested are processed.

**Comment by mrshenli** [comment]
@lw @pritamdamania87 @jjlilley @xush6528 

Any thoughts on how we should handle the exceptions thrown by a user callback? We used to catch and handle those errors in the callbacks we inserted, but as we are going to expose this to Python API, we can no longer guarantee noexcept callbacks.

Pythons `concurrent.futures.Future` does the following.

> Added callables are called in the order that they were added and are always called in a thread belonging to the process that added them. If the callable raises an Exception subclass, it will be logged and ignored. If the callable raises a BaseException subclass, the behavior is undefined.

We could do the same, or maybe let `add_done_callback` to return another `Future` to propagate the error?

cc @rohan-varma

**Comment by jjlilley** [comment]
The current future impls don't really catch/handle the exceptions in a great manner - they seem to mostly just blow up in the stack of whoever called markCompleted(). And in practice, prevent the remaining callbacks from being executed (maybe we should just called that "undefined behavior")

In the rpc code, the strategy seems to be mostly around avoiding throwing in the callbacks.

The Python strategy of explicitly logging and ignoring doesn't sound so bad at all.

An alternative would be to do something more like we see in folly::Future<>, though that impl feels somewhat more complex/obfuscated, and perhaps not what we want either.

**Comment by rohan-varma** [comment]
Agreed with @jjlilley, following the python protocol here seems to make sense. It could be a pain to debug if the user has several layers of callbacks added with `add_done_callback()`, and having these throw outside of the main thread may result in hard to debug issues.

**Comment by rohan-varma** [comment]
@mrshenli About this block of code:

```
future_ret = rpc.Future()
    # nested RPC calls
    future_value = rpc.rpc_async(dst2, my_add_2, args=(a, b))
    # finish processing asynchronously when the nested RPC is done.
    future_value.add_callback(
        lambda ret: future_ret.mark_complete(ret + c)
    )
    return future_ret
```

If this pattern is going to be common, should we have a `.then()` API that returns a future that is satisfied when the callback is completed? Such as:

```
future_value = rpc.rpc_async(dst2, my_add_2, args=(a, b))
future_ret = future_value.then(func)
return future_ret
```
The API will take the return value of `future_value`, call `func` on it, and then mark future_ret complete with the result of that computation. It saves the extra lines of creating the future and marking it as complete. It'll also allow chaining as follows:
```
fut = rpc.rpc_async(...).then(asyncfn1).then(asyncfn2)...
```
(though not sure if this is great programming style)
Since this is also useful for RPC profiling, I added a separate issue to track: https://github.com/pytorch/pytorch/issues/37908

**Comment by mrshenli** [comment]
@rohan-varma 

This indeed looks good. And with this, we can implement the first version of async UDF without worrying about which package the `Future` should belong. 

Eventually, we would still need to expose the `Future` ctor as the async part can come from non-RPC code, e.g. file IO, batch processing, etc.

**Comment by mrshenli** [comment]
@wanchaol @xush6528 

If we follow the folly Future's design, the callback in the `then(cb)` API takes the return value instead of the Future, and the type of the return value can be anything. Will this be a problem if we later want to use `future.then(cb)` in a TorchScript function?

**Comment by wanchaol** [comment]
> @wanchaol @xush6528
> 
> If we follow the folly Future's design, the callback in the `then(cb)` API takes the return value instead of the Future, and the type of the return value can be anything. Will this be a problem if we later want to use `future.then(cb)` in a TorchScript function?

@mrshenli We will have to add bindings to TorchScript for either `then` or `add_done_callback` when this needs to be used in TorchScript. 

For the return value can be `anything`, we should ensure those bindings that if this is getting executed under TorchScript, the `cb` must be a TorchScript function, otherwise error out to the user.

**Comment by mrshenli** [comment]
@wanchaol 

Does this mean as long as `cb` is a `TorchScript` function and we have bindings for `then`/`add_done_callback`, it would work. Curious, is this because JIT only need to know that `then(cb)`'s `cb` arg is a function and does not need `cb` to stick with a fixed signature? Or do we need to add multiple bindings for, e.g., `then(cb(x: Tensor))` and `then(cb(x: str))`?

**Comment by wanchaol** [comment]
> @wanchaol
> 
> Does this mean as long as `cb` is a `TorchScript` function and we have bindings for `then`/`add_done_callback`, it would work. Curious, is this because JIT only need to know that `then(cb)`'s `cb` arg is a function and does not need `cb` to stick with a fixed signature? Or do we need to add multiple bindings for, e.g., `then(cb(x: Tensor))` and `then(cb(x: str))`?

It would need TorchScript to fully support first class functions(right now we don't yet), but we could do it a similar way with `torch.jit._fork` where it needs only a TorchScript function, as we could infer argument types from the function's schema.

**Comment by mrshenli** [comment]
closed by #39216 #39486 #39485 #39267
